torch/csrc/api/include/torch/nn/module.h

```diff
@@ -113,25 +113,39 @@ class CloneableModule : public Module {
  public:
   using Module::Module;
 
-  // should it also detach the gradients, like a deepcopy? Or maybe let's just
-  // give clone() a boolean for this?
-  std::unique_ptr<Module> clone() const override {
-    auto ptr = std::unique_ptr<Module>(
-        new Derived(*static_cast<const Derived*>(this)));
+  virtual void reset() = 0;
+
+  std::shared_ptr<Derived> build() {
+    auto module = std::make_shared<Derived>(static_cast<Derived&&>(*this));
+    module->reset();
+    return std::move(module);
+  }
+
+  std::shared_ptr<Module> clone() const override {
+    auto ptr = std::make_shared<Derived>(*static_cast<const Derived*>(this));
+    ptr->parameters_.clear();
+    ptr->reset();
     for (auto& parameter : ptr->parameters_) {
-      parameter.second = this->parameters_.at(parameter.first).clone();
+      parameter.second.data().copy_(
+          this->parameters_.at(parameter.first).data());
     }
     for (auto& child : ptr->children_) {
       child.second = this->children_.at(child.first)->clone();
     }
     return ptr;
   }
 };
-
-template <class Module>
-std::shared_ptr<typename std::decay<Module>::type> make(Module&& module) {
-  auto ptr = std::make_shared<typename std::decay<Module>::type>(
-      std::forward<Module>(module));
-  return ptr;
-}
 }} // namespace torch::nn
+
+#define TORCH_PARAMETER(T, name)             \
+ public:                                     \
+  auto name(T new_##name)->decltype(*this) { \
```

Comment:
nit: can you add spaces around `->` so that it doesn't look as if you're dereferencing a pointer?

Reply:
clang-format does this for some reason